### PR TITLE
controller: Proactively disconnect node based on heartbeat

### LINKED ISSUE #857: RFE: periodic heartbeat in bluechi-controller
## Please describe what you would like to see

In https://github.com/eclipse-bluechi/bluechi/issues/652 the problem was raised that `bluechi-controller` detects a disconnect of the `bluechi-agent` quite late if the cable is unplugged, for example, and a command is issued to that agent before the connection timeout is hit. This leads to a "zombie" agent in the `bluechi-controller` - it still lists the agent as online and refuses reconnects of that agent (due to the name still being used). 
This can be mitigated via the introduced TCP KeepAlive options (https://github.com/eclipse-bluechi/bluechi/issues/674), but it still takes quite a while to detect it (depending on various complex tcp options, e.g. tcp retransmissions). 

The `bluechi-agent` on the other side can detect a disconnect rather soon due to the Heartbeat feature. Such a periodic check of the connection status on an application layer could be used in the `bluechi-controller` as well. Based on the last seen timestamp, it could actively disconnect nodes. 

**Note:** This only makes sense for rather reliable networks, I think, and should be deactivated by default (so no overhead).

 
## Please describe the solution you'd like

The `bluechi-controller` uses the same event-based mechanism that is used in the `bluechi-agent` for the heartbeat to check in a configurable interval for each node that is online the last seen timestamp is not older than a configurable threshold. If it is older, it actively disconnects the node. 

- New configuration options for `bluechi-controller`
   - [x] `HeartbeatInterval`: The interval for checking the last seen timestamps of nodes in milliseconds, a value of 0 disables it. 
   - [x] `NodeHeartbeatThreshold`: If `now - last_seen_timestamp > NodeHeartbeatThreshold` then actively disconnect the node
- [x] Implement verify and disconnect logic
- [x] Implement integration tests
- [x] Extend documentation (man pages, examples, etc.)

### PR DESCRIPTION
**Pull request by dofmind**
This PR is an update of https://github.com/eclipse-bluechi/bluechi/pull/870

The main difference is that the code is simplified by using microseconds in last_seen instead of the struct timespec.
It updates last_seen in node_method_register() to prevent the node disconnecting right now when last_seen is 0 or too old.
Additionally, it handles disconnected nodes by calling node_disconnected() instead of controller_remove_node(). Otherwise an error will occur.
```
bluechi-controller: ../git/src/controller/node.c:123: unit_subscriptions_clear: Assertion `LIST_IS_EMPTY(usubs->subs)' failed.
```

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/857

### CODE REVIEW

File: doc/man/bluechi-controller.conf.5.md [hunk 1]
```diff
@@ -27,6 +27,14 @@ A comma separated list of unique bluechi-agent names. It's mandatory to set the
 in the list can connect to `bluechi-controller'. These names are defined in the agent's configuration file under `NodeName`
 option (see `bluechi-agent.conf(5)`).
 
+#### **HeartbeatInterval** (long)
+
+The interval to periodically check node's connectivity based on heartbeat signals sent to bluechi, in milliseconds.
```
Comment (engelmi) [inline doc/man/bluechi-controller.conf.5.md]: I think it would be good to mention here that a value of `0` disables the periodic check:
```suggestion
The interval to periodically check node's connectivity based on heartbeat signals sent to bluechi, in milliseconds. A value of 0 disables it. 
```

Not part of this PR, but this should be added to the description of the [Agents Heartbeat](https://github.com/eclipse-bluechi/bluechi/blob/main/config/agent/agent.conf#L29) as well.

File: doc/man/bluechi-controller.conf.5.md [hunk 2]
```diff
@@ -27,6 +27,14 @@ A comma separated list of unique bluechi-agent names. It's mandatory to set the
...
+
+#### **NodeHeartbeatThreshold** (long)
+
+The threshold in milliseconds to determine whether a node is disconnected. If the node's last heartbeat signal was received before this threshold, bluechi treats that the node has been disconnected.
```
Comment (engelmi) [inline doc/man/bluechi-controller.conf.5.md]:
```suggestion
The threshold in milliseconds to determine whether a node is disconnected. If the node's last heartbeat signal was received before this threshold, bluechi assumes that the node is down or the connection was cut off and performs a disconnect.
```

File: src/controller/controller.c [hunk 3]
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
...
+                }
+
+                if (now < node->last_seen) {
+                        bc_log_error("Clock skew detected");
+                        return -EINVAL;
+                }
+
+                diff = now - node->last_seen;
+                if (diff > (uint64_t) controller->heartbeat_threshold_msec * USEC_PER_MSEC) {
+                        node_disconnect(node);
```
Comment (engelmi) [inline src/controller/controller.c:516]: It would be good to have an info log here to provide a bit of information why a node gets disconnected:
```suggestion
                       bc_log_infof("Did not receive heartbeat from node '%s' since '%d'ms. Disconnecting it...", node->name, controller->heartbeat_threshold_msec);
                        node_disconnect(node);
```

File: src/controller/controller.c [hunk 4]
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
...
+                uint64_t diff = 0;
+                uint64_t now = 0;
+
+                if (!node_is_online(node)) {
+                        continue;
+                }
+
+                now = get_time_micros();
+                if (now == 0) {
+                        return -EINVAL;
```
Comment (engelmi) [inline src/controller/controller.c]: Probably worth adding a log here:
```suggestion
                       bc_log_error("Failed to get the time");
                        return -EINVAL;
```

Also, the `return`ing here will result in `controller_reset_heartbeat_timer` not being called - so this would completely stop controller heartbeat. Is this desired?

File: src/controller/controller.c [hunk 5]
```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
...
+
+static int controller_setup_heartbeat_timer(Controller *controller) {
+        _cleanup_(sd_event_source_unrefp) sd_event_source *event_source = NULL;
+        int r = 0;
+
+        assert(controller);
+
+        if (controller->heartbeat_interval_msec <= 0) {
+                bc_log_warnf("Heartbeat disabled since interval is '%d', ",
+                             controller->heartbeat_interval_msec);
```
Comment (engelmi) [inline src/controller/controller.c]:
```suggestion
                bc_log_warnf("Heartbeat disabled since configured interval '%d' is <=0",
                             controller->heartbeat_interval_msec);
```

Again, not related to the PR, but could you change this as well [for the agent here](https://github.com/eclipse-bluechi/bluechi/blob/main/src/agent/agent.c#L190).

File: config/controller/controller.conf [hunk 6]
```diff
@@ -16,6 +16,14 @@
 # The port the controller listens on to establish connections with the bluechi agents.
 #ControllerPort=842
 
+#
+# Defines the interval between two heartbeat signals sent to bluechi in milliseconds.
```
Comment (engelmi) [inline config/controller/controller.conf]:
```suggestion
# Defines the interval for which the bluechi-controller periodically checks the connectivity of all nodes based on the received heartbeat signals sent to bluechi-agent to bluechi-controller, in milliseconds.
```

File: config/controller/controller.conf [hunk 7]
```diff
@@ -16,6 +16,14 @@
...
+#HeartbeatInterval=2000
+
+#
+# Defines the threshold to actively disconnect nodes based on the time difference from the last heartbeat signal.
```
Comment (engelmi) [inline config/controller/controller.conf]:
```suggestion
# Defines the threshold to actively disconnect nodes based on the time difference from the last received heartbeat signal from the respective bluechi-agent. In milliseconds.
```

File: config/controller/controller.conf [hunk 8]
```diff
@@ -16,6 +16,14 @@
 # The port the controller listens on to establish connections with the bluechi agents.
 #ControllerPort=842
 
+#
+# Defines the interval between two heartbeat signals sent to bluechi in milliseconds.
+#HeartbeatInterval=2000
```
Comment (engelmi) [inline config/controller/controller.conf]: The values here should reflect the default ones. 

For `HeartbeatInterval` I think a default value of `0` would be good to disable this feature by default. 
The default value of `6000` for the `NodeHeartbeatThreshold` seems reasonable to me. 

Also, we decided to hard-code the defaults in and set these before parsing any configuration files. Therefore, please add 
- the default values also in the [protocol.h](https://github.com/eclipse-bluechi/bluechi/blob/main/src/libbluechi/common/protocol.h#L29) next to the other defaults. 
- setting these hard-coded values in the [cfg_controller_def_conf](https://github.com/eclipse-bluechi/bluechi/blob/main/src/libbluechi/common/cfg.c#L490-L511)

**Review by engelmi — COMMENTED** [review]
Thanks for your contribution! @dofmind 

Overall the PR looks good. I added a few minor comments, mostly about logging. I tested your changes manually (also via valgrind) and it works well. Nonetheless, please add two integration tests:
- checking that `HeartbeatInteral=0` disables it 
(e.g. by confirming the agent stays connected after x sec when both, agent and controller heartbeat are disabled and controller threshold is y sec)
- checking that the agent gets disconnected when not the threshold is reached (e.g. by disabling the agent heartbeat)

File: src/controller/controller.c (hunk 4 above)
Comment (dofmind) [reply src/controller/controller.c:502]: Okay, let's handle this with a continue statement rather than a return statement.

Comment (dofmind) [reply src/controller/controller.c]: Okay, let's handle this with a continue statement rather than a return statement.

File: doc/man/bluechi-controller.conf.5.md (hunk 1 above)
Comment (dofmind) [reply doc/man/bluechi-controller.conf.5.md]: done

File: doc/man/bluechi-controller.conf.5.md (hunk 2 above)
Comment (dofmind) [reply doc/man/bluechi-controller.conf.5.md]: done

File: src/controller/controller.c (hunk 3 above)
Comment (dofmind) [reply src/controller/controller.c:516]: done

File: src/controller/controller.c (hunk 5 above)
Comment (dofmind) [reply src/controller/controller.c]: done

File: config/controller/controller.conf (hunk 6 above)
Comment (dofmind) [reply config/controller/controller.conf]: done

File: config/controller/controller.conf (hunk 7 above)
Comment (dofmind) [reply config/controller/controller.conf]: done

File: config/controller/controller.conf (hunk 8 above)
Comment (dofmind) [reply config/controller/controller.conf]: done

File: doc/man/bluechi-controller.conf.5.md (hunk 1 above)
Comment (engelmi) [reply doc/man/bluechi-controller.conf.5.md]: Thanks!

**Comment by engelmi** [comment]
LGTM and thanks for adding tests!

@dofmind Could you add the two new configuration options to the [bluechi-controller.conf.5.md](https://github.com/eclipse-bluechi/bluechi/blob/main/doc/man/bluechi-controller.conf.5.md)? Sorry, I missed that during the previous review. Then, I think, the PR is ready to merge.

**Comment by dofmind** [comment]
> @dofmind Could you add the two new configuration options to the [bluechi-controller.conf.5.md](https://github.com/eclipse-bluechi/bluechi/blob/main/doc/man/bluechi-controller.conf.5.md)? Sorry, I missed that during the previous review. Then, I think, the PR is ready to merge.

This PR already has them. Did I miss anything?

**Comment by engelmi** [comment]
> > @dofmind Could you add the two new configuration options to the [bluechi-controller.conf.5.md](https://github.com/eclipse-bluechi/bluechi/blob/main/doc/man/bluechi-controller.conf.5.md)? Sorry, I missed that during the previous review. Then, I think, the PR is ready to merge.
> 
> This PR already has them. Did I miss anything?

No, but I did. Sorry about that.

**Review by engelmi — APPROVED** [review]
LGTM

Created an additional documentation issue for extending the readthedocs https://github.com/eclipse-bluechi/bluechi/issues/913
